### PR TITLE
8165276: Spec states to invoke the premain method in an agent class if it's public but implementation differs

### PR DESCRIPTION
**Pull request by sspitsyn**
This change have been already reviewed by Mandy, Sundar, Alan and David.
Please, see the jdk 15 review thread:
  http://mail.openjdk.java.net/pipermail/serviceability-dev/2020-June/031998.html

Now, the PR approval is needed.
The push was postponed because the CSR was not approved at that time (it is now):
   https://bugs.openjdk.java.net/browse/JDK-8248189
Investigation of existing popular java agents was requested by Joe.
----------

The java.lang.instrument spec:
  https://docs.oracle.com/en/java/javase/15/docs/api/java.instrument/java/lang/instrument/package-summary.html

Summary:
  The java.lang.instrument spec clearly states:
    "The agent class must implement a public static premain method
     similar in principle to the main application entry point."
  Current implementation of sun/instrument/InstrumentationImpl.java
  allows the premain method be non-public which violates the spec.
  This fix aligns the implementation with the spec.

Testing:
  A mach5 run of jdk_instrument tests is in progress.

---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed

### Issue
 * [JDK-8165276](https://bugs.openjdk.java.net/browse/JDK-8165276): Spec states to invoke the premain method in an agent class if it's public but implementation differs

### Reviewers
 * [Mandy Chung](https://openjdk.java.net/census#mchung) (@mlchung - **Reviewer**) ⚠️ Review applies to 219a2cd681819b6b73e8d115362d1d4846e8f66f
 * [David Holmes](https://openjdk.java.net/census#dholmes) (@dholmes-ora - **Reviewer**)
 * [Alan Bateman](https://openjdk.java.net/census#alanb) (@AlanBateman - **Reviewer**) ⚠️ Review applies to 219a2cd681819b6b73e8d115362d1d4846e8f66f

### Download
`$ git fetch https://git.openjdk.java.net/jdk pull/1694/head:pull/1694`
`$ git checkout pull/1694`

### CODE REVIEW

File: src/java.instrument/share/classes/sun/instrument/InstrumentationImpl.java
Comment (AlanBateman) [inline src/java.instrument/share/classes/sun/instrument/InstrumentationImpl.java:501]: If canAccess fails then it means that javaAgentClass is not public or the premain method is not public. The invoke will fail but I agree eat explicit canAccess check means we get finer control on the exception message.

(I can't help feeling that we should do a bit more cleanup here and not use getDeclaredMethod or be concerned with inheritance. This is because the Premain-Class is meant to specify the agent class. Also can't have a public static premain in super class and a non-public static premain in a sub-class).

**Comment by plummercj** [comment]
> This change have been already reviewed by Mandy, Sundar, Alan and David.
> Now, the PR approval is needed.

Can you provide a link to the discussion? I'm mostly curious if there was some discussion as to why Instrument purposefully allowed non-public premain methods:

```507         // the premain method should not be required to be public,
508         // make it accessible so we can call it
509         // Note: The spec says the following:
510         //     The agent class must implement a public static premain method...
511         setAccessible(m, true);```

**Comment by AlanBateman** [comment]
All the discussion is in the bug and CSR:
https://bugs.openjdk.java.net/browse/JDK-8165276
https://bugs.openjdk.java.net/browse/JDK-8248189
We messed up in JDK-5070281 (JDK 6) and it came to light in JDK 9 when auditing the use of setAccessible in the JDK.

**Comment by sspitsyn** [comment]
Chris, I've added link to the jdk 15 review thread to the PR description.

File: src/java.instrument/share/classes/sun/instrument/InstrumentationImpl.java
Comment (sspitsyn) [reply src/java.instrument/share/classes/sun/instrument/InstrumentationImpl.java:501]: My gut feeling is that it should be possible to get rid of the getDeclaredMethod calls with the reasoning you provided above. The canAccess check is not needed in such a case as the getMethod returns public methods only (is my understanding correct?). This would simplify the implementation. Two disadvantages of this approach are:
- less fine control on the exception message: NoSuchMethodException instead of IllegalAccessException
- some compatibility risk is involved (most of agents define premain method correctly, so the only java agents that define premain methods with unusual tricks are impacted)

Comment (mlchung) [reply src/java.instrument/share/classes/sun/instrument/InstrumentationImpl.java:501]: If the java agent class declares a non-public premain method but  its superclass declares a public premain method, what should be the expected behavior?   The proposed fix will choose the java agent class's non-public premain and therefore it will fail to load the agent class with IAE.

If we get rid of the `getDeclaredMethod` call and find premain using `getMethod`, it will choose the public premain method defined directly in the java agent class or indirectly in its superclass.

As Alan stated, `Premain-Class` attribute is meant to specify the agent class.  Also specified in the package spec of `java.lang.instrument', the agent class must implement a public static premain method similar in principle to the main application entry point.  Also under "Manifest Attributes" section,
```
   Premain-Class
        When an agent is specified at JVM launch time this attribute specifies the agent class. That is, the class containing the premain method. 
```
 It expects that there is a public premain method declared in the agent class, not in its superclass.

So I think it's best to fix this as well in this PR by dropping line 469-495 (i.e. keep `getDeclaredMethod` case and not search for inherited premain).   Throw if the premain method is not found or it is not accessible.   This needs to update the CSR for behavioral change.

**Comment by sspitsyn** [comment]
Alan, Mandy and Chris,
Thank you for your comments!
I've committed and pushed an update which implements the suggestion from Mandy to get rid of inherited premain/agentmain methods. I also had to fix 3 tests (InheritAgent0100, InheritAgent1000 and InheritAgent1100) which started to fail with this update as they are relying on the inherited premain methods. I've refactored them to negative tests.

File: src/java.instrument/share/classes/sun/instrument/InstrumentationImpl.java
Comment (mlchung) [inline src/java.instrument/share/classes/sun/instrument/InstrumentationImpl.java:499]: The comments in line 429-443 need to be updated.

**Review by mlchung — COMMENTED** [review]
Please update @bug in the tests to include this bug ID.

All InheritAgentXXX tests are updated to have the InheritAgentXXX class public.  However, I think you need to go through them individually for the new test cases per this behavioral change.    For example InheritAgent0100 previously verifies the invocation of the premain in its superclass.  Does the modified test ensure that this fails to load?  Per the comment in the new InheritAgent0100Super class, it expects the superclass' premain should be called.

**Comment by sspitsyn** [comment]
Mandy, thank you for comments.
You are right, the comments at 429-443 tests need some updates. I'm working on it now.

**Comment by sspitsyn** [comment]
I've updated the **@bug** in the tests to include this bug ID.
All the comments from Mandy have been addressed now.

File: src/java.instrument/share/classes/sun/instrument/InstrumentationImpl.java
Comment (AlanBateman) [inline src/java.instrument/share/classes/sun/instrument/InstrumentationImpl.java:468]: I think the updated patch looks much better. 

If the agent class doesn't declare a premain then NoSuchMethodException will be thrown.

The only thing that I'm wondering about now is the exception message when the agent class is not public. If I read the changes correctly it means that IllegalAccessException will be thrown saying that <class>.premain should be public. Is that correct? We might need to adjust to the exception message to make it clear, or put in an explicit then to ensure to test that the agent class is public.

Comment (dholmes-ora) [reply src/java.instrument/share/classes/sun/instrument/InstrumentationImpl.java:468]: I think an explicit check for a public premain method is better to disambiguate the cases and provide appropriate error messages.

**Review by dholmes-ora — CHANGES_REQUESTED** [review]

**Comment by sspitsyn** [comment]
Alan, David and Many, thank you for the comments!
I'll prepare an update according to the recent requests from you.
One question is if I need to clone this PR to the JDK 16 fork:
  https://github.com/openjdk/jdk16
It depends on our chances to finalize this before RDP2.
An alternate approach would be to continue reviewing this PR until all comment are resolved and then clone it to jdk16.

**Comment by sspitsyn** [comment]
Mandy, thank you for the suggestion.
I'll retarget the bug and CSR to jdk 17 as nobody is objecting.

Also, wanted to make it clear about Exception messages that are provided for two different cases.

The test java/lang/instrument/NonPublicPremainAgent.java defines a non-public premain method:
`    static void premain(String agentArgs, Instrumentation inst) {`

With the m.canAccess check the following message is provided (it looks right):
`  Exception in thread "main" java.lang.IllegalAccessException: method NonPublicPremainAgent.premain must be declared public `

Without this check the message was (not sure, it is good enough):
`  Exception in thread "main" java.lang.IllegalAccessException: class sun.instrument.InstrumentationImpl (in module java.instrument) cannot access a member of class NonPublicPremainAgent with modifiers "static"`

Also, I've added a new test java/lang/instrument/NonPublicAgent.java which defines a non-public agent class with a public premain method.

With the m.canAccess check the following message is provided (it does not look right):
`  Exception in thread "main" java.lang.IllegalAccessException: method NonPublicPremainAgent.premain must be declared public `

Without this check the message is (it looks pretty confusing):
`  Exception in thread "main" java.lang.IllegalAccessException: class sun.instrument.InstrumentationImpl (in module java.instrument) cannot access a member of class NonPublicAgent with modifiers "public static"`

So, it seems we also need an explicit check for agent class being public with a right message provided.
Please, let me know what you think.

I've done some tests refactoring motivated by some private requests from Mandy. Will publish the update as soon as it is ready. Hopefully, today.

**Comment by dholmes-ora** [comment]
The agent class doesn't have to be public it just has to be accessible.

The premain method should be queried for public modifier rather than just relying on a failed invocation request.

**Comment by sspitsyn** [comment]
David, thank you for catching this. I'm probably missing something here.
If the agent class is not public then the `m.canAccess(null)` check is not passed and IAE is thrown with the message:
  `Exception in thread "main" java.lang.IllegalAccessException: method NonPublicAgent.premain must be declared public`

But the `NonPublicAgent.premain` is declared public as below:
```
    public static void premain(String agentArgs, Instrumentation inst) {
        System.out.println("premain: NonPublicAgent was loaded");
    }
```
It seems, the IAE is thrown because the agent class is not public.
Does it mean the `m.canAccess(null)` check is not fully correct?

**Comment by sspitsyn** [comment]
Thanks, David!
Yes, I was also thinking that the setAccessible has to be remained after all the checks.

**Comment by sspitsyn** [comment]
I've pushed an update with the following changes:
- InstrumentationImpl.java update with the process sequence suggested by David. More specifically: the premain/agentmain method public modifier is checked instead of m.canAccess, after that if the agent class is not public then the setAccessible is called to make the method invokeable;
- The tests are refactored (tried to implement suggestion from Mandy to simplify the tests). It includes:
   - new helper class are added to build agent jar file:  `test/jdk/java/lang/instrument/AgentJarBuilder.java`
   - new helper class to run negative tests which are expected to throw an exception: `test/jdk/java/lang/instrument/NegativeAgentRunner.java`
   - introduced in the first push new shell script is removed: `test/jdk/java/lang/instrument/PremainClass/MakeJAR.sh`
   - the test runners with names *Test.java are removed, all the tests use jtreg commands instead (negative tests now use the NegativeAgentRunner)
   - the test  `test/jdk/java/lang/instrument/NonPublicPremainAgent.java` is moved to the PremainClass sub-folder

It might make sense to remove previously added public modifiers to several agent classes. However, I've decided to skip it for now. Please, let me know if you think it is desired thing to do.
Mandy also suggested to consider using the exception message format produced by thrown by `jdk.internal.reflect.Reflection::newIllegalAccessException`. It looks like this: `Exception in thread "main" java.lang.IllegalAccessException: class sun.instrument.InstrumentationImpl (in module java.instrument) cannot access a member of class NonPublicAgent with modifiers "public static"`. Please, let me know if this format would be better.

**Comment by AlanBateman** [comment]
Mandy is correct, the bigger picture here is accessibility. Code in the java.instrument module should be able to invoke the premain or agentmain without needing to suppress access checks. It's a perquisite to supporting the deployment of java agents as modules. When an agent is deployed as a module it will need the agent class to be public in a package that is exported to at least the java.instrument module. The premain method will need to be public too.
So PR1694 is really just about aligning the spec so that the premain method is public. However, it does not fix the overall accessibility issue. Fixing the accessibility issue will fixing the java.lang.instrument package description to specify that the agent class is public.

File: test/jdk/java/lang/instrument/NegativeAgentRunner.java
Comment (mlchung) [inline test/jdk/java/lang/instrument/NegativeAgentRunner.java:42]: This should also check if the exit value is non-zero.

File: test/jdk/java/lang/instrument/AgentJarBuilder.java
Comment (mlchung) [inline test/jdk/java/lang/instrument/AgentJarBuilder.java:36]: Can you use the test library `test/lib/jdk/test/lib/util/JavaAgentBuilder.java`?

File: test/jdk/java/lang/instrument/PremainClass/InheritAgent0010.java
Comment (mlchung) [inline test/jdk/java/lang/instrument/PremainClass/InheritAgent0010.java:32]: I don't see `sun.tools.jar` is used.  Is this qualified exports needed?

**Comment by mlchung** [comment]
> So PR1694 is really just about aligning the spec so that the premain method is public. However, it does not fix the overall accessibility issue.

OK I see that JDK-5070281 changes to allow the agent class to be non-public as the main class that declares `public static void main`.   So `setAccessible` should only be called if the agent class is in an unnamed module.  

If the agent class is in a named module and not accessible to java.instrument module (if the java agent is packaged in a modular JAR file), it should fail with IAE - can you confirm?    I wanted to understand the current behavior.

**Comment by sspitsyn** [comment]
Mandy, I've pushed updates with changes:
- NegativeAgentRunner checks if the exit value is non-zero
- AgentJarBuilder is replaced with JavaAgentBuilder
- removed unneeded qualified exports

> So setAccessible should only be called if the agent class is in an unnamed module.
Now the agent class is always loaded as in the unnamed module.
I tried to test this with a modular jar file which has module-info.class in it.
In fact, I don't know if there are any options that would help to load the premain agent class in named module.
I guess, it has to be implemented as part of the enhancement https://bugs.openjdk.java.net/browse/JDK-6932391 .

The following patch can be committed if you like as it does not harm:
@@ -476,11 +476,13 @@ public class InstrumentationImpl implements Instrumentation {
             String msg = "method " + classname + "." +  methodname + " must be declared public";
             throw new IllegalAccessException(msg);
         }
-
-        if ((javaAgentClass.getModifiers() & java.lang.reflect.Modifier.PUBLIC) == 0) {
-            // If the java agent class is not public, make the premain/agentmain
-            // accessible, so we can call it.
+        if ((javaAgentClass.getModifiers() & java.lang.reflect.Modifier.PUBLIC) == 0 &&
+            !javaAgentClass.getModule().isNamed()) {
+            // If the java agent class is not public and is in unnamed module
+            // then make the premain/agentmain accessible, so we can call it.
             setAccessible(m, true);
         }

File: src/java.instrument/share/classes/sun/instrument/InstrumentationImpl.java
Comment (mlchung) [inline src/java.instrument/share/classes/sun/instrument/InstrumentationImpl.java:475]: An alternative way: `if (Modifier.isPublic(m.getModifiers()))`

Similar can be applied for checking if the javaagentClass is public.

**Review by mlchung — COMMENTED** [review]
@sspitsyn  Thanks for the update.   I think it's good to add the unnamed module check before calling `setAccessible` that becomes clear that this check only intends for unnamed module.   Maybe make it clear in the comment something like this: 

```
    // if the java agent class is in an unnamed module, the java agent class can be non-public.
    // suppress access check upon the invocation of the premain/agentmain method
```

Since the agent class can be non-public, it means that it's not necessary to modify the test agent class to public as you did in this patch.

I don't think `@library /test` is needed, isn't it?  The test library classes are under test/lib.

**Comment by sspitsyn** [comment]
@mlchung Thank you for the suggestions. I've addressed them in the latest update.
The `@library /test` is needed only in the tests that use NegativeAgentRunner as this class is located in folder one level up.
I've removed the this line from the other tests.

**Review by mlchung — APPROVED** [review]
The change looks okay with minor comments on the tests.    Nit: it will be good to update `@summary` in the InheritAgentxxxx tests that now fail to load the agent.  

W.r.t. the premain/agentmain method declared in the agent class, the specification clearly states that in the package summary:

> The agent class must implement a public static premain method similar in principle to the main application entry point.

> The agent class must implement a public static agentmain method.

**Comment by sspitsyn** [comment]
@mlchung All suggestions are right and addressed in the latest update.

File: test/jdk/java/lang/instrument/RetransformAgent.java
Comment (mlchung) [inline test/jdk/java/lang/instrument/RetransformAgent.java:26]: The copyright header and `@bug` should be reverted.

File: test/jdk/java/lang/instrument/SimpleAgent.java
Comment (mlchung) [inline test/jdk/java/lang/instrument/SimpleAgent.java:2]: The copyright header change should be reverted.

**Comment by sspitsyn** [comment]
@mlchung
Thank you for the catch! Fixed it.
Also, reverted the fix in the NativeMethodPrefixAgent.java.
Additionally, got rid of the 'inherited' term in the summary of InheritAgent*.java tests.

**Review by mlchung — APPROVED** [review]
Thanks for the change.  Looks okay.

**Comment by sspitsyn** [comment]
Thank you a lot, Mandy!
Sorry for missed changes and overlooked issues in tests.

**Review by AlanBateman — APPROVED** [review]
Thanks for persevering on this one. The updated checks look good. At some point I think we re-examine the issue of allowing agents to be non-public, maybe emit a warning and eventually disallow it. If/when support is added for developing and deployed java agents as modules it might be the time to re-examine it.

**Comment by sspitsyn** [comment]
Thank you for review, Alan!

**Comment by sspitsyn** [comment]
This PR has been approved by Mandy and Alan.
I'm also waiting for approval from David and the CSR approval from Joe.

**Review by dholmes-ora — APPROVED** [review]
Hi Serguei,

This seems good to go.

Thanks,
David

**Comment by sspitsyn** [comment]
Thank you for review, David!

**Pull request closed without merging**
